zebra-consensus/src/block/subsidy/funding_streams.rs

```diff
@@ -48,49 +48,34 @@ pub fn funding_stream_values(
     Ok(results)
 }
 
-/// Returns the address change period
-/// as described in [protocol specification §7.10][7.10]
-///
-/// [7.10]: https://zips.z.cash/protocol/protocol.pdf#fundingstreams
-fn funding_stream_address_period(height: Height, network: &Network) -> u32 {
-    // Spec equation: `address_period = floor((height - (height_for_halving(1) - post_blossom_halving_interval))/funding_stream_address_change_interval)`,
-    // <https://zips.z.cash/protocol/protocol.pdf#fundingstreams>
-    //
-    // Note that the brackets make it so the post blossom halving interval is added to the total.
-    //
-    // In Rust, "integer division rounds towards zero":
-    // <https://doc.rust-lang.org/stable/reference/expressions/operator-expr.html#arithmetic-and-logical-binary-operators>
-    //   This is the same as `floor()`, because these numbers are all positive.
-
-    let height_after_first_halving = height - network.height_for_first_halving();
-
-    let address_period = (height_after_first_halving + POST_BLOSSOM_HALVING_INTERVAL)
-        / FUNDING_STREAM_ADDRESS_CHANGE_INTERVAL;
-
-    address_period
-        .try_into()
-        .expect("all values are positive and smaller than the input height")
-}
-
 /// Returns the position in the address slice for each funding stream
 /// as described in [protocol specification §7.10][7.10]
 ///
 /// [7.10]: https://zips.z.cash/protocol/protocol.pdf#fundingstreams
 fn funding_stream_address_index(height: Height, network: &Network) -> usize {
-    let num_addresses = network.num_funding_streams();
+    let funding_streams = network.funding_streams(height);
 
     let index = 1u32
         .checked_add(funding_stream_address_period(height, network))
         .expect("no overflow should happen in this sum")
         .checked_sub(funding_stream_address_period(
-            FUNDING_STREAM_HEIGHT_RANGES
-                .get(&network.kind())
-                .unwrap()
-                .start,
+            funding_streams.height_range().start,
             network,
         ))
         .expect("no overflow should happen in this sub") as usize;
 
+    // Funding stream recipients may not have the same number of addresses on configured Testnets,
+    // the number of addresses for each recipient should be validated for a configured height range
+    // when configured Testnet parameters are built.
+    let num_addresses = funding_streams
+        .recipients()
+        .values()
+        .next()
+        // TODO: Return an Option from this function and replace `.unwrap()` with `?`
+        .unwrap()
+        .addresses()
+        .len();
```

Comment:
I'm a bit confused by this code. What is the purpose of the `next()` call?

Reply:
It's to get the first item in the recipients `HashMap`, any of them would be fine since they should all be the same length on Mainnet and Testnet, or they should all be checked when testnet::Parameters is being built on configured Testnets.

Reply:
Got it. I think this might return the wrong thing since the first element could be the Deferred recipient, whose address length is 0, which will lead to a panic in the assert below

Reply:
Thank you! Fixed in https://github.com/ZcashFoundation/zebra/pull/8694/commits/7df2e2863f75c2b85323e8906f9f30ee0bacc73f.